Add Close, Fanout and RollbackAndForward to our MBT

v0d1ch opened the pull request:
In order to improve our MBT `Close`, `Fanout` and `RollbackAndForward` are added as the actions. Now the `MockChain` is only rolling forward and only rolls back when appropriate action is produced.

There is also new test case _fanout contains whole confirmed UTxO_ that should further assure us that the `Fanout` `UTxO` matches with whatever we transacted in the open `Head`.

NOTE:
Currently we are not able to compare equality between `Model` representation of the `UTxO` (`[(CardanoSigningKey, Value)]`) with the real `UTxO` . Even if the value and the address matches between the two the input `Ix` we generate when converting starts always from zero and we can't just sort `TxOuts` and compare that since there is no `Ord` instance.

This is related to #1260 as it will help reproduce this bug and preparatory work for #1057 

---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

v0d1ch committed:
Model Close and Fanout and check fannout utxo

v0d1ch committed:
Add fixed era horizon

v0d1ch committed:
Add missing precondition

v0d1ch committed:
All parties should see fanout

v0d1ch committed:
Revert changes to the contestation period generation

v0d1ch committed:
Apply hints

v0d1ch committed:
Retry Fanout action

v0d1ch committed:
Remove redundant actions and implement correctly check for the fanout

There was a change of plans for the postcondition check so I removed
CheckFanoutUTxO and GetConfirmedUTxO since we can check directly
the Fanout utxo against the result in postcondition.

v0d1ch committed:
Introduce RollbackAndForward action

Instead of doing rollbacks automatically in the MockChain we modelled
the Action to do the same from the Model.

v0d1ch committed:
Just compare the values for now

We are not able to check equality between a real UTxO and the
Model representation. For now we just check that the Fanout
value matches.

v0d1ch commented on hydra-node/test/Hydra/Model.hs:
```diff
@@ -353,7 +375,7 @@ genToCommit (hk, ck) = do
 genContestationPeriod :: Gen ContestationPeriod
 genContestationPeriod = do
-  n <- choose (1, 10)
+  n <- choose (1, 200)
```
These are the values that were actually present before so I just reverted the changes I did in a previous pr.

v0d1ch commented on hydra-node/test/Hydra/Model.hs:
```diff
@@ -464,21 +486,25 @@ instance
+            decorateFailure st action (toTxOuts finalUTxO) (toList result)
+            let getLovelace = selectLovelace . foldMap txOutValue
+            -- TODO: Is it enough to just compare the values? Should we make
+            -- sure there is equal number of outputs and that we can actually
```
What I mean is: is it enough to just compare the values are matching. What if we end up in a situation where we fanout one output where we should have actually created two or more.

v0d1ch replied:
Yeah, that I could do.

v0d1ch commented on hydra-node/test/Hydra/ModelSpec.hs:
```diff
@@ -176,6 +176,34 @@ spec = do
+      tx <- action $ Model.NewTx party payment
+      eventually (ObserveConfirmedTx tx)
+      action_ $ Model.Close party
+      void $ action $ Model.Fanout party
```
It is a bit hard to do that since what we get here is wrapped in a `Var` and also the `tx` is `Payment` so it needs to be converted into `UTxO` and then somehow compared or find another workaround.

ch1bo replied:
This is what we would have wanted, but Quvic people said that this is currently not possible. The alternative would be an assertive action `AssertFanoutUTxO` or so, which then would do the check in the `postcondition` (or `perform`). We ended up doing this in all `Fanout` actions.

A comment should be added here though that the check is actually happening in the postcodition.

v0d1ch committed:
Some enhansements from the review

v0d1ch committed:
Compare list of outputs sorted by lovelace on Fanout postcondition

v0d1ch committed:
Format

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/TimeHandle.hs:
```diff
@@ -70,6 +70,21 @@ instance Arbitrary TimeHandle where
+      -- in the future
+      safeZone = 3 * 216000 / 0.05
+      horizonSlot = SlotNo $ truncate $ uptimeSeconds + safeZone
+  pure $ mkTimeHandle currentSlotNo (SystemStart startTime) (eraHistoryWithHorizonAt horizonSlot)
```
Should: create a time handle without a horizon

Instead of using a horizon "far in the future" (what is far enough?), we should be able to create an era history containing of a singleton era without an end.

Also, this (or the function I suggest) should not be in the "production code" but somewhere closer to the tests (or in the `testlib`), so we don't use it in production by accident.

ch1bo commented on hydra-node/test/Hydra/Model.hs:
```diff
@@ -197,6 +187,12 @@ instance StateModel WorldState where
         frequency
           [ (5, genNewTx)
           , (1, genClose)
+          , (1, genRollbackAndForward)
```
Could: also generate rollback and forward in the `Initial`state.

ch1bo commented on hydra-node/test/Hydra/Model.hs:
```diff
@@ -217,6 +213,16 @@ instance StateModel WorldState where
+      -- TODO: investigate why choosing higher number of blocks e.g. (1, 5)
+      -- causes arithmetic underflow. Maybe we need a pointer to the current
+      -- block so we don't rollback past that one?
+      numberOfBlocks <- choose (1, 2)
```
Most likely because the mock chain does not guard against `rollbackAndForward` with too high numbers ("not so many blocks to rollback")

v0d1ch replied:
True. I added the check in the MockChain.

ch1bo commented on hydra-node/test/Hydra/Model.hs:
```diff
@@ -235,6 +241,16 @@ instance StateModel WorldState where
+      Initial{} -> False
+      Idle{} -> False
+      Start{} -> False
+      Final{} -> False
```
Should: this action be available in all states.. no?

ch1bo commented on hydra-node/test/Hydra/Model.hs:
```diff
@@ -464,21 +487,24 @@ instance
+      Fanout{} ->
+        case hydraState st of
+          Final{finalUTxO} -> do
+            let sortByLovelace = sort . fmap (selectLovelace . txOutValue)
```
Should: not reduce to lovelace values, but compare the full `[TxOut]`

We can use `sortOn` to achieve that. An even better sorting would be to use both, the address and the value to order. For example:

```suggestion
            let sortByLovelace = sortOn (\o -> (txOutAddress o, txOutValue o))
```

ch1bo replied:
Should: document why this is needed

ch1bo commented on hydra-node/test/Hydra/Model.hs:
```diff
@@ -512,7 +540,9 @@ instance
           case find headIsOpen outputs of
             Just _ -> pure ()
-            Nothing -> error "The head is not open for node"
+            Nothing -> failure "The head is not open for node"
```
```suggestion
            Nothing -> error "The head is not open for node"
```

I think `error` provides a better output (it's ironic, but the model test suite does not utilize the more detailed `HUnitFailure`)

ch1bo commented on hydra-node/test/Hydra/Model.hs:
```diff
@@ -605,6 +634,16 @@ performCommit parties party paymentUTxO = do
+
+-- | NOTE: This function generates input 'Ix' that start from zero so
+-- can't be used with certainty when want to check equality with some 'UTxO'
+-- even though the value and the key would match.
```
Could: fix this function to do that

Would be an alternative to `toTxOuts` + the sorting. A nice property here would be:

```
propIsDistributive :: Arbitrary a => (a -> b) -> [a] -> [a] -> Property
propIsDistributive fn as bs =
  fn as <> fn bs === fn (as <> bs)
```

with the test being `prop "realUTxO is distributive" $ propIsDistributive toRealUTxO`

(I hope this was distributivity .. long time no math)

ch1bo commented on hydra-node/test/Hydra/Model.hs:
```diff
@@ -696,6 +743,33 @@ performClose party = do
+performFanout party = do
+  nodes <- gets nodes
+  let thisNode = nodes ! party
+  waitForClosed thisNode
```
Should: not wait for closed, but for `ReadyToFanout`

Don't we see model failures where we try to fanout "too early" if we only wait for `HeadIsClosed` message?

ch1bo commented on hydra-node/test/Hydra/Model.hs:
```diff
@@ -696,6 +743,33 @@ performClose party = do
+  outputs <- lift $ serverOutputs thisNode
+  case find headIsFinalized outputs of
+    Just HeadIsFinalized{utxo} -> pure utxo
+    _ -> failure "Failed to perform Fanout"
```
Should: not look twice for `HeadIsFinalized`, but already parse `utxo` from the first `waitUntilMatch` (might need to enhance this function)

ch1bo commented on hydra-node/test/Hydra/ModelSpec.hs:
```diff
@@ -176,6 +177,26 @@ spec = do
+      tx <- action $ Model.NewTx party payment
+      eventually (ObserveConfirmedTx tx)
+      action_ $ Model.Close party
+      void $ action $ Model.Fanout party
```
```suggestion
      action_ $ Model.Fanout party
```

ch1bo reviewed (approved):
Good if it works, but a bit surprising that we don't need to wait for `ReadyToFanout`?

Also consider my "Should" suggestions, please.

v0d1ch committed:
Move and rename fixed TimeHandle

Also use indefinite Horizon

v0d1ch committed:
Add a check for the number of blocks to rollback

Add this check to the MockChain and generate more blocks
in the Model

v0d1ch committed:
Wait for ReadyToFanout on Fanout action

v0d1ch committed:
Generate Rollback on Init too

Fix precondition to always be able to rollback

v0d1ch committed:
Do not rollback on init and genereate smaller rollback blocks

Currently the rollback in the Init can easily go back past the
current block so for now we can ommit it. Check in the MockChain
is needed to make this proper since the one I added does not work.

v0d1ch committed:
Don't wait twice on Fanout

Just wait to see the appropriate server output or throw
the error after some time.

v0d1ch committed:
Sort by address and value

v0d1ch committed:
Format

v0d1ch committed:
Fix headOpensIfAllPartiesCommit

Why did I have to do this fix? This needs investigation

v0d1ch committed:
Add pending test related to distributive property of toRealUTxO
